Add Big Endian support to Pure Java Snappy

kiwi1969 opened the pull request:
Added code to allow Pure Java Snappy to work on BIG_ENDIAN systems
This allows snappy to work on a wider range of systems such as z/OS, where the native DLL required for JNI call may not exist

xerial commented on src/main/java/org/xerial/snappy/pure/UnsafeUtil.java:
```diff
@@ -66,4 +73,54 @@ public static long getAddress(Buffer buffer)
+ 
+    public static short getShort(Object inputBase, long offset)
+    {
+    	if (order.equals(ByteOrder.LITTLE_ENDIAN))
```
Thank you for the workaround, but checking the endianness for each getXXX operation has too much overhead for a performance-sensitive library like snappy-java. Can we do this check only once and switch the implementations for little-endian and big-endian at startup time?

kiwi1969 replied:
I will see if I can do something

kiwi1969 replied:
OK, I have an alternate solution which is a bit better
I created 2 new methods SnappyRawCompressorBE and SnappyRawDecompressorBE which have the reversebytes in.
This way the LITTLE_ENDIAN code is intact.
I instead I now have IF statements in PureSnappyJava.java, to select the normal or the BE method for each Compress/Decompress method

kiwi1969 replied:
OK, last update. I made a new SnappyAPI interface, so there are no IF statements needed, other than on initial Loader.
So there should be no performance issues for old flows, and no net change to any of the existing modules, apart from the Loader. See 86b8578

kiwi1969 replied:
OK, rebased the 'net unchanged' code back to xerial base, as trivial whitespaces/tabs etc came out as changes, to make diff report come out clean.

xerial commented:
ok. The code change is fairly large, but basically it looks like adding BE version of PureJavaSnappy. 

Unlike the concerns discussed in airlift/aircompressor#116, the pura-java support is just a fallback when no native library is built, so merging this PR should be ok if we can test running it somewhere. 

@kiwi1969 Any thought? Can you run ./sbt test on your big-endian platform?

zhengxiaomei123 commented:
Hi , once the PR is merged, Will this package be upload to maven central repo? How to use this code in other project such as presto? Thanks.

kiwi1969 commented:
Yes, only difference is the loader conditionally loads the alternative BE class, which uses alternative BE subclasses.
The only difference in those classes, is reversebytes method happening in some of the places used for encoding/decoding. (I didn't need the reversebytes for the simple memory copies, so is better than my original code)

Note the 'fallback' currently doesn't happen automatically, as you throw an error, which stops the code from falling into else statement in snappy-loader.
Only way I could get to the pure was via the JVM parm

My employer doesn't have any Java CI build/test system on Z/OS, so I just currently compile on my Mac, and copied the resulting jar files to Z/OS, then just ran a very simple test case (which worked, and encoded a UTF string exactly same byte code as on Mac, and decoded correctly).
I will have a look at ./sbt test to see what other test cases I can run.

xerial commented:
> Note the 'fallback' currently doesn't happen automatically
```try ... catch(Exception e) { ... } ``` didn't work? We may need to switch to `catch(Throwable e)`.

kiwi1969 commented:
For CI, I see you already have a Linux s390x test running. I am not sure how the previous 'Snappy Pure' tests listed in the logs are currently working, as s390x is Big Endian and will fail with "Big Endian not supported" error.
Maybe you could just tweak what you have there now, and double check that the libsnappy.so is not being used.

xerial commented:
@kiwi1969 Ah. That's right. s390x is a big-endian machine. Travis CI provides an emulation environment for s390x.

Then adding this line to `.travis.yml` can be used for testing BE version of this PR:
```
 - ./sbt -Dorg.xerial.snappy.purejava=true "testOnly *Pure*"
```

xerial commented on src/main/java/org/xerial/snappy/SnappyLoader.java:
```diff
@@ -170,7 +177,10 @@ static synchronized SnappyApi loadSnappyApi()
         catch(Exception e) {
             // Fall-back to pure-java Snappy implementation
-            setSnappyApi(new PureJavaSnappy());
+            if (order.equals(ByteOrder.LITTLE_ENDIAN))
+        		setSnappyApi(new PureJavaSnappy());
+        	else
+        		setSnappyApi(new PureJavaSnappyBE());
```
This is a duplicated code. We can extract this part, for example, as `private static void loadPureJavaSnappy() { ... }`.

kiwi1969 replied:
I actually found that for AirCompressor snappy testing, the overhead of just doing the littleEndian conversion in specific places was insignificant.
Do we still want to have 2 sets of classes or just do the test in specific places in the RawCompress/RawDecompress methods ?

xerial commented:
A weird thing is if s390x emulation on Travis is a big-endian, it should have thrown an error because aircompressor doesn't support it. But no error was shown.

kiwi1969 commented:
I swapped to the version that matches what I uploaded to Aircompressor, as the overhead really wasn't that much. Having issue with build system. I don't think it is me

kiwi1969 commented:
I have found that my work project utilizing Apache Avro with snappy compress option, is working very well with my patches to  snappy-java-1.1.7.7 on Z/OS (when I add snappy-java-1.1.7.7 to Maven .pom file) and specify PureJava JVM directive for snappy.

Note: Rather than do a complete rebuild, we just compiled the 3 class files, and replaced them directly in the .jar (aka zipfile).
The other application jar files then did not notice the difference, and did not need rebuilding (which would be needed if I changed the name of the jar)
Anyone needing a quick fix until officially merged can do the same.

xerial commented:
Thanks for the update. Let me take a look later so that we can merge this PR and ship a new version

xerial commented:
OK. The code became simple enough and we can expect JIT optimization will work for such a switch for the endian check. I'll merge this PR. Thanks!
